openedx-unsupported/devstack
# invert dependencies so that backends now depend on frontends (plus some cleanup)

## Description

The intent of this PR is to resolve an immediate developer experience issue: _since the Learning MFE is now enabled by default, the command `make dev.up.lms` on its own is no longer sufficient to run the default Open edX courseware experience_.

We hope that the fix for this issue will help us resolve a deeper, upcoming developer experience issue: _soon [^1], several micro-frontends will need to be started alongside backend services in order to provide a working Open edX experience out-of-the-box_.

Changes are best reviewed by-commit.

1. The first commit is some doc fixes related to devserver restarting.
2. The next commit is a superficial-yet-breaking interface change in order to make frontend app names consistent in Devstack. We are doing this clean-up opportunistically, since we'll need to announce workflow-breaking changes from this PR anyway.
3. The next commit is a non-user-facing style cleanup (some YAML alphabetization).
4. The next commit is a non-user-facing refactor (pulling XQueue into the main docker-compose.yml file).
5. The penultimate commit is a major interface change, along with an ADR. In short, backend services will be defined as depending on frontend services, such that `make dev.up.<backend-service>` automatically starts all required frontends service.
6. The final commit makes it so local NPM packages can be used with devstack microfrontends. There is an included ADR, and I am [making frontend-build doc updates here](https://github.com/edx/frontend-build/pull/193).

I will communicate all interface changes via email and Slack.

[^1]: When? Not sure; it's up to individual squads that own MFEs. ~We do know that Engage squad has work in-flight to make the Course Home features run in the Learning MFE by default.~ (update: they just did!) Also, several MFEs run as the default in Lilac, so it would make sense if those were made to run in Devstack by default as well.

## Supporting information

[TNL-8407](https://openedx.atlassian.net/browse/TNL-8407)

Further details are in the included ADRs (0004 and 0005).

Related PR to update frontend-build docs: https://github.com/edx/frontend-build/pull/193

## Testing instructions

Suggested smoke test for service dependencies:

* Ensure that `make dev.up.lms` brings up at least LMS, frontend-app-gradebook and frontend-app-learning. Ensure that those services are usable.
* Ensure that `make dev.attach.frontend-app-gradebook` attaches to the Gradebook MFE container.
* Ensure that `make dev.up.xqueue` brings up XQueue and MySQL 5.7.
* Ensure that `make dev.down` brings down all services.

Suggested smoke test for local package mounting:

* Start LMS and fa-learning with `make lms-up`.
* Load the demo course, confirming that it loads in the Learning MFE using the default (bootstrap blue) Open edX theme.
* Clone https://github.com/edx/brand-edx.org into the `src/` directory within your devstack workspace.
* Run `npm install && npm build` in that repo.
* Follow [these update docs](https://github.com/edx/frontend-build/blob/kdmccormick/local-package-docs/README.rst#local-module-configuration-for-webpack), copying `module.config.js` into frontend-app-learning.
* In `module.config.js`, set `@edx/brand` to the directory `../src/brand-edx.org`. Comment out all other local modules.
* Shell into frontend-app-learning (`make frontend-app-learning-shell`) and run `npm install`. Exit shell.
* Restart frontend-app-learning.
* Load the demo course again, confirming that it shows the edX.org elm-green theme.
